### PR TITLE
Fix odo update

### PR DESCRIPTION
**Pull request by girishramnani**
## What is the purpose of this change? What does it change?
Use of --context flag in odo update and internally odo update uses common push.

## Was the change discussed in an issue?
fixes https://github.com/openshift/odo/issues/2155
fixes https://github.com/openshift/odo/issues/1778
fixes https://github.com/openshift/odo/issues/1782

for the issues https://github.com/openshift/odo/issues/1782 and https://github.com/openshift/odo/issues/1778. I have uncommented the tests and haven't observed the flake since locally and on CI.

## How to test changes?

```
odo create nodejs
odo push
odo update --git  https://github.com/openshift/nodejs-ex.git

### CODE REVIEW

**Comment by girishramnani** [comment]
haven't tested heavily for all scenarios hence WIP.
Also I think update related tests are commented.

File: pkg/odo/cli/component/update.go [hunk 1]
```diff
@@ -26,44 +24,38 @@ const UpdateRecommendedCommandName = "update"
...
 		`)
 
 // NewUpdateOptions returns new instance of UpdateOptions
 func NewUpdateOptions() *UpdateOptions {
-	return &UpdateOptions{ComponentOptions: &ComponentOptions{}}
+	return &UpdateOptions{
+		CommonPushOptions: &CommonPushOptions{
+			pushConfig: true, // we assume if someone updates then the config is only pushed
+			show:       false,
+			forceBuild: false,
```
Comment (girishramnani) [inline pkg/odo/cli/component/update.go]: @kadel should we force build everytime someone updates?

**Comment by kadel** [comment]
@girishramnani any progress on this?

**Comment by girishramnani** [comment]
terraform error
```
level=fatal msg="failed to fetch Terraform Variables: failed to fetch dependency of \"Terraform Variables\": failed to fetch dependency of \"Bootstrap Ignition Config\": failed to fetch dependency of \"Master Machines\": failed to generate asset \"Platform Credentials Check\": validate AWS credentials: checking install permissions: error simulating policy: Throttling: Rate exceeded\n\tstatus code: 400, request id: 66a3f9b6-2c11-420d-82f4-1613a7b4e866"
2020/01/02 09:27:57 Container setup in pod integration-e2e-benchmark failed, exit code 1, reason Error
Another process exited
```
/retest

**Comment by girishramnani** [comment]
```
Pulling image "registry.svc.ci.openshift.org/ci/clonerefs@sha256:0121c9382a8d31174e6c580ad1b6fcbe06c5d4dad6a7e02c02e3dfd058c2fba0" ...
error: error pulling image registry.svc.ci.openshift.org/ci/clonerefs@sha256:0121c9382a8d31174e6c580ad1b6fcbe06c5d4dad6a7e02c02e3dfd058c2fba0: Get https://registry.svc.ci.openshift.org/v2/ci/clonerefs/manifests/sha256:0121c9382a8d31174e6c580ad1b6fcbe06c5d4dad6a7e02c02e3dfd058c2fba0: Get https://registry.svc.ci.openshift.org/openshift/token?account=serviceaccount&scope=repository%3Aci%2Fclonerefs%3Apull: net/http: TLS handshake timeout
2020/01/02 12:35:20 No custom metadata found and prow metadata already exists. Not updating the metadata.
```
/retest

**Comment by girishramnani** [comment]
again same error
```
Pulling image "registry.svc.ci.openshift.org/ci/clonerefs@sha256:0121c9382a8d31174e6c580ad1b6fcbe06c5d4dad6a7e02c02e3dfd058c2fba0" ...
error: error pulling image registry.svc.ci.openshift.org/ci/clonerefs@sha256:0121c9382a8d31174e6c580ad1b6fcbe06c5d4dad6a7e02c02e3dfd058c2fba0: Get https://registry.svc.ci.openshift.org/v2/ci/clonerefs/manifests/sha256:0121c9382a8d31174e6c580ad1b6fcbe06c5d4dad6a7e02c02e3dfd058c2fba0: Get https://registry.svc.ci.openshift.org/openshift/token?account=serviceaccount&scope=repository%3Aci%2Fclonerefs%3Apull: net/http: TLS handshake timeout
```

**Comment by girishramnani** [comment]
```
Pulling image "registry.svc.ci.openshift.org/ci/clonerefs@sha256:0121c9382a8d31174e6c580ad1b6fcbe06c5d4dad6a7e02c02e3dfd058c2fba0" ...
error: error pulling image registry.svc.ci.openshift.org/ci/clonerefs@sha256:0121c9382a8d31174e6c580ad1b6fcbe06c5d4dad6a7e02c02e3dfd058c2fba0: Get https://registry.svc.ci.openshift.org/v2/ci/clonerefs/manifests/sha256:0121c9382a8d31174e6c580ad1b6fcbe06c5d4dad6a7e02c02e3dfd058c2fba0: Get https://registry.svc.ci.openshift.org/openshift/token?account=serviceaccount&scope=repository%3Aci%2Fclonerefs%3Apull: net/http: TLS handshake timeout
```
/retest

**Comment by girishramnani** [comment]
If you observe that these are all `if` statements and not `if.. elseif .. else` because the conditions are not mutually exclusive. We want all the if statements to execute because there is a `checkFlag` variable that increases if either of the 3 flags `git`, `local` or `binary` is provided. This checkFlag is then used to make sure that only 1 of all these flags are provided.

**Comment by girishramnani** [comment]
So we cannot use switch case to replace all the if statements

**Comment by girishramnani** [comment]
self approving for now.
/approve

**Comment by dharmit** [comment]
I think we're missing a check that confirms the existence of the directory when using `odo update --local <dir-name>`. Here's the error/panic I'm seeing when `./code` doesn't exist:

```sh

$ mkcd nodejs-ex

$ git clone https://github.com/sclorg/nodejs-ex .
Cloning into '.'...
remote: Enumerating objects: 631, done.
remote: Total 631 (delta 0), reused 0 (delta 0), pack-reused 631
Receiving objects: 100% (631/631), 235.11 KiB | 431.00 KiB/s, done.
Resolving deltas: 100% (243/243), done.

$ odo create nodejs node
 ✓  Validating component [6ms]

Please use `odo push` command to create the component with source deployed

$ odo update --local ./code

Configuration changes
 ✓  Initializing component
 ✗  Creating component [76933ns]
 ✗  Failed to create component with name node. Please use `odo config view` to view settings used to create component. Error: stat ./code: no such file or directory
failed to get info of path ./code of component {Name:node SourcePath:./code SourceRef: SourceType:local ImageName:nodejs EnvVars:[DEBUG_PORT=5858] Ports:[8080/TCP] Resources:&ResourceRequirements{Limits:ResourceList{},Requests:ResourceList{},} ApplicationName:app Wait:false StorageToBeMounted:map[] StdOut:<nil>}
github.com/openshift/odo/pkg/component.CreateComponent
        /home/dshah/go/src/github.com/openshift/odo/pkg/component/component.go:413
github.com/openshift/odo/pkg/odo/cli/component.(*CommonPushOptions).createCmpIfNotExistsAndApplyCmpConfig
        /home/dshah/go/src/github.com/openshift/odo/pkg/odo/cli/component/common_push.go:72
github.com/openshift/odo/pkg/odo/cli/component.(*CommonPushOptions).Push
        /home/dshah/go/src/github.com/openshift/odo/pkg/odo/cli/component/common_push.go:145
github.com/openshift/odo/pkg/odo/cli/component.(*UpdateOptions).Run
        /home/dshah/go/src/github.com/openshift/odo/pkg/odo/cli/component/update.go:139
github.com/openshift/odo/pkg/odo/genericclioptions.GenericRun
        /home/dshah/go/src/github.com/openshift/odo/pkg/odo/genericclioptions/runnable.go:28
github.com/openshift/odo/pkg/odo/cli/component.NewCmdUpdate.func1
        /home/dshah/go/src/github.com/openshift/odo/pkg/odo/cli/component/update.go:159
github.com/openshift/odo/vendor/github.com/spf13/cobra.(*Command).execute
        /home/dshah/go/src/github.com/openshift/odo/vendor/github.com/spf13/cobra/command.go:702
github.com/openshift/odo/vendor/github.com/spf13/cobra.(*Command).ExecuteC
        /home/dshah/go/src/github.com/openshift/odo/vendor/github.com/spf13/cobra/command.go:783
github.com/openshift/odo/vendor/github.com/spf13/cobra.(*Command).Execute
        /home/dshah/go/src/github.com/openshift/odo/vendor/github.com/spf13/cobra/command.go:736
main.main
        /home/dshah/go/src/github.com/openshift/odo/cmd/odo/odo.go:80
runtime.main
        /usr/local/go/src/runtime/proc.go:203
runtime.goexit
        /usr/local/go/src/runtime/asm_amd64.s:1357

# there's no directory name 'code' in the pwd
$ ls
helm  openshift  tests  views  package.json  README.md  server.js
```

File: pkg/odo/cli/component/update.go [hunk 2]
```diff
@@ -73,16 +66,43 @@ func (uo *UpdateOptions) Complete(name string, cmd *cobra.Command, args []string
 
 // Validate validates the update parameters
 func (uo *UpdateOptions) Validate() (err error) {
+
+	uo.doesComponentExist, err = component.Exists(uo.Context.Client, uo.localConfigInfo.GetName(), uo.localConfigInfo.GetApplication())
+	if err != nil {
+		return errors.Wrapf(err, "failed to check if component of name %s exists in application %s", uo.localConfigInfo.GetName(), uo.localConfigInfo.GetApplication())
```
Comment (dharmit) [inline pkg/odo/cli/component/update.go]: I'm wondering when this error is thrown. I did below but didn't face this error:
1. Created a component with `odo create`
2. Before pushing it, I did `odo update`.

So in this case, component doesn't exist in the application yet, right? At least not on the OpenShift cluster. Am I right in expecting this error in this scenario? :thinking:

Comment (girishramnani) [reply pkg/odo/cli/component/update.go]: actually if the component doesn't exist then `doesComponentExist` become false but the error is thrown when maybe connection to the cluster timeout or we dont have permissions to check if the component exists or not

**Comment by girishramnani** [comment]
> I think we're missing a check that confirms the existence of the directory when using `odo update --local <dir-name>`. Here's the error/panic I'm seeing when `./code` doesn't exist:
> 
> ```shell
> $ mkcd nodejs-ex
> 
> $ git clone https://github.com/sclorg/nodejs-ex .
> Cloning into '.'...
> remote: Enumerating objects: 631, done.
> remote: Total 631 (delta 0), reused 0 (delta 0), pack-reused 631
> Receiving objects: 100% (631/631), 235.11 KiB | 431.00 KiB/s, done.
> Resolving deltas: 100% (243/243), done.
> 
> $ odo create nodejs node
>  ✓  Validating component [6ms]
> 
> Please use `odo push` command to create the component with source deployed
> 
> $ odo update --local ./code
> 
> Configuration changes
>  ✓  Initializing component
>  ✗  Creating component [76933ns]
>  ✗  Failed to create component with name node. Please use `odo config view` to view settings used to create component. Error: stat ./code: no such file or directory
> failed to get info of path ./code of component {Name:node SourcePath:./code SourceRef: SourceType:local ImageName:nodejs EnvVars:[DEBUG_PORT=5858] Ports:[8080/TCP] Resources:&ResourceRequirements{Limits:ResourceList{},Requests:ResourceList{},} ApplicationName:app Wait:false StorageToBeMounted:map[] StdOut:<nil>}
> github.com/openshift/odo/pkg/component.CreateComponent
>         /home/dshah/go/src/github.com/openshift/odo/pkg/component/component.go:413
> github.com/openshift/odo/pkg/odo/cli/component.(*CommonPushOptions).createCmpIfNotExistsAndApplyCmpConfig
>         /home/dshah/go/src/github.com/openshift/odo/pkg/odo/cli/component/common_push.go:72
> github.com/openshift/odo/pkg/odo/cli/component.(*CommonPushOptions).Push
>         /home/dshah/go/src/github.com/openshift/odo/pkg/odo/cli/component/common_push.go:145
> github.com/openshift/odo/pkg/odo/cli/component.(*UpdateOptions).Run
>         /home/dshah/go/src/github.com/openshift/odo/pkg/odo/cli/component/update.go:139
> github.com/openshift/odo/pkg/odo/genericclioptions.GenericRun
>         /home/dshah/go/src/github.com/openshift/odo/pkg/odo/genericclioptions/runnable.go:28
> github.com/openshift/odo/pkg/odo/cli/component.NewCmdUpdate.func1
>         /home/dshah/go/src/github.com/openshift/odo/pkg/odo/cli/component/update.go:159
> github.com/openshift/odo/vendor/github.com/spf13/cobra.(*Command).execute
>         /home/dshah/go/src/github.com/openshift/odo/vendor/github.com/spf13/cobra/command.go:702
> github.com/openshift/odo/vendor/github.com/spf13/cobra.(*Command).ExecuteC
>         /home/dshah/go/src/github.com/openshift/odo/vendor/github.com/spf13/cobra/command.go:783
> github.com/openshift/odo/vendor/github.com/spf13/cobra.(*Command).Execute
>         /home/dshah/go/src/github.com/openshift/odo/vendor/github.com/spf13/cobra/command.go:736
> main.main
>         /home/dshah/go/src/github.com/openshift/odo/cmd/odo/odo.go:80
> runtime.main
>         /usr/local/go/src/runtime/proc.go:203
> runtime.goexit
>         /usr/local/go/src/runtime/asm_amd64.s:1357
> 
> # there's no directory name 'code' in the pwd
> $ ls
> helm  openshift  tests  views  package.json  README.md  server.js
> ```

if you observe the stack trace closely its not happening due to a panic. This is an intentional behaviour of the `errors.Wrapf` function. It returns a stack trace as well when someone formats the error with `%+v`.
```
// Wrapf returns an error annotating err with a stack trace
// at the point Wrapf is call, and the format specifier.
// If err is nil, Wrapf returns nil.
func Wrapf(err error, format string, args ...interface{}) error {
	if err == nil {
		return nil
	}
	err = &withMessage{
		cause: err,
		msg:   fmt.Sprintf(format, args...),
	}
	return &withStack{
		err,
		callers(),
	}
}
```

**Comment by girishramnani** [comment]
But definitely we should check if the source location exists or not in the `Validate` function of update itself

**Comment by girishramnani** [comment]
so found the place where the stack trace got originated. its due to the `%+v` present on this line
https://github.com/openshift/odo/blob/fa4a47c8a1c3b0d00799758e20bf733a966c0e77/pkg/odo/cli/component/common_push.go#L74

File: pkg/odo/cli/component/update.go (hunk 2 above)
Comment (dharmit) [reply pkg/odo/cli/component/update.go]: Makes sense about the `false` thing. But I still don't hit this error when I stopped my crc cluster. Instead, after building from the latest bits, I get the no such directory error which is covered in the `Validate` function but after this particular piece:
```sh
$ odo update --local ./code
 ✗  error while validating source path: stat ./code: no such file or directory
```

Comment (girishramnani) [reply pkg/odo/cli/component/update.go]: Not sure how to forcefully get that error :sweat_smile:

**Pull request merged**